src/FEValues/cell_values.jl

```diff
@@ -34,137 +34,222 @@ utilizes scalar shape functions and `CellVectorValues` utilizes vectorial shape
+# FIXME sdim should be something like `length(value(geo_interpol))`
+function CellScalarValues(valtype::Type{T}, quad_rule::QuadratureRule, func_interpol::Interpolation,
+        geo_interpol::Interpolation=func_interpol; sdim::Int=getdim(func_interpol))  where {T}
+    CellScalarValues(valtype, quad_rule, func_interpol, geo_interpol, Val(sdim))
+end
+
+function CellScalarValues(::Type{T}, quad_rule::QuadratureRule{rdim,shape}, func_interpol::Interpolation{rdim,shape},
+        geo_interpol::Interpolation{rdim,shape}, ::Val{sdim}) where {rdim,T,shape<:AbstractRefShape,sdim}
 
-    @assert getdim(func_interpol) == getdim(geom_interpol)
-    @assert getrefshape(func_interpol) == getrefshape(geom_interpol) == shape
     n_qpoints = length(getweights(quad_rule))
 
     # Function interpolation
     n_func_basefuncs = getnbasefunctions(func_interpol)
     N    = fill(zero(T)          * T(NaN), n_func_basefuncs, n_qpoints)
-    dNdx = fill(zero(Vec{dim,T}) * T(NaN), n_func_basefuncs, n_qpoints)
-    dNdξ = fill(zero(Vec{dim,T}) * T(NaN), n_func_basefuncs, n_qpoints)
+    dNdx = fill(zero(SVector{sdim,T}) * T(NaN), n_func_basefuncs, n_qpoints)
+    dNdξ = fill(zero(SVector{rdim,T}) * T(NaN), n_func_basefuncs, n_qpoints)
 
     # Geometry interpolation
-    n_geom_basefuncs = getnbasefunctions(geom_interpol)
+    n_geom_basefuncs = getnbasefunctions(geo_interpol)
     M    = fill(zero(T)          * T(NaN), n_geom_basefuncs, n_qpoints)
-    dMdξ = fill(zero(Vec{dim,T}) * T(NaN), n_geom_basefuncs, n_qpoints)
+    dMdξ = fill(zero(SVector{rdim,T}) * T(NaN), n_geom_basefuncs, n_qpoints)
 
     for (qp, ξ) in enumerate(quad_rule.points)
         for i in 1:n_func_basefuncs
             dNdξ[i, qp], N[i, qp] = gradient(ξ -> value(func_interpol, i, ξ), ξ, :all)
         end
         for i in 1:n_geom_basefuncs
-            dMdξ[i, qp], M[i, qp] = gradient(ξ -> value(geom_interpol, i, ξ), ξ, :all)
+            dMdξ[i, qp], M[i, qp] = gradient(ξ -> value(geo_interpol, i, ξ), ξ, :all)
         end
     end
 
     detJdV = fill(T(NaN), n_qpoints)
 
-    CellScalarValues{dim,T,shape}(N, dNdx, dNdξ, detJdV, M, dMdξ, quad_rule, func_interpol, geom_interpol)
+    CellScalarValues{sdim,rdim,T,shape}(N, dNdx, dNdξ, detJdV, M, dMdξ, quad_rule, func_interpol, geo_interpol)
 end
 
 # CellVectorValues
-struct CellVectorValues{dim,T<:Real,refshape<:AbstractRefShape,M} <: CellValues{dim,T,refshape}
-    N::Matrix{Vec{dim,T}}
-    dNdx::Matrix{Tensor{2,dim,T,M}}
-    dNdξ::Matrix{Tensor{2,dim,T,M}}
+#   rdim = reference element dimension
+#   sdim = spatial dimension
+#   vdim = vector dimension (i.e. dimension of evaluation of what `value` should return)
+#   M1   = number of elements in the matrix dNdx (should be vdim × sdim)
+#   M2   = number of elements in the matrix dNdξ (should be vdim × rdim)
+struct CellVectorValues{sdim,rdim,T<:Real,refshape<:AbstractRefShape,vdim,M1,M2} <: CellValues{sdim,rdim,T,refshape}
```

Comment:
Parameter ordering here should also be thought about I think. All of them are needed here, but the ones you would dispatch on should move to the front so you don't have to write `CellVectorValues{<:Any, <:Any, <:Any, <:Any, TheThingIReallyWant}`.

Reply:
Perhaps `CellVectorValues{vdim, sdim, rdim, T, refshape, M1, M2}` makes sense? Then the first parameter the dimension for  `function_value`, the first two parameters for `function_gradient`. Probably `T` is more useful than `rdim`, but looks nicer to group all the dims together :)

We should also remove `refshape` and replace it with the interpolation(s) as I suggested in https://github.com/Ferrite-FEM/Ferrite.jl/pull/674#discussion_r1161686800 so that you can dispatch on that for selecting the mapping, for example (this part can/should be done in a separate PR though).

Reply:
I mean we should be able to query reference dimension and the reference geometry via `Interpolation` anyway. Will fix.

To elaborate on the ordering, I have put vdim on the end because this keeps the parametrization order the same between the values. But I think it also makes sense. To order them by importance as you suggest.

Reply:
If you write a method for both of them you will dispatch on `CellValues` though, so then the order of the subtypes doesn't matter.